python/lsst/ts/genericcamera/genericcameracsc.py

```diff
@@ -455,23 +570,162 @@ async def liveView_loop(self, exposure_time):
         self.isLive = False
         self.log.info("liveView_loop - End")
 
+    async def run_auto_exposure_loop(self, min_exp_time, max_exp_time, configuration):
+        """Run the auto exposure capture loop.
+
+        The cadence of the images is determined by the value of
+        AUTO_EXP_IMAGE_INTERVAL.
+
+        Parameters
+        ----------
+        min_exp_time: `float`
+            The minimum exposure time to use.
+        max_exp_time: `float`
+            The maximum exposure time to use.
+        configuration: `str`
+            A Yaml string containing additional configuration
+            parameters.
+        """
+        self.log.info("autoExposure_loop - Start")
+        self.isAutoExposure = True
+
+        shutter = False
+        sensors = ""
+        keyValueMap = ""
+        obsNote = ""
+        if configuration != "":
+            config = yaml.safe_load(configuration)
+            if "shutter" in config:
+                shutter = config["shutter"]
+            if "sensors" in config:
+                sensors = config["sensors"]
+            if "keyValueMap" in config:
+                keyValueMap = config["keyValueMap"]
+            if "obsNote" in config:
+                obsNote = config["obsNote"]
+
+        images_in_sequence = 0
+        image_index = 0
+        exposure_time = min_exp_time
+
+        try:
+            while self.runAutoExposureTask:
+                timestamp = time.time()
+                image_name = self.fileNameFormat.format(
+                    timestamp=int(timestamp),
+                    index=image_index,
+                    total=images_in_sequence,
+                )
+
+                self.evt_startTakeImage.put()
+                await self.camera.startTakeImage(
+                    exposure_time,
+                    shutter,
+                    sensors,
+                    keyValueMap,
+                    obsNote,
+                )
+                exposure = None
+
+                background_level = 0.0
+                while not (
```

Comment:
I think this loop inside a loop is were you are determining the exposure time. Consider having this in a separate method, (e.g. `determine_auto_exposure_time`). In fact I think you may need two different methods; `determine_auto_exposure_time` and `adjust_auto_exposure_time`.

`determine_auto_exposure_time` will start from scratch, evaluate images until it gets a good exposure time. 

`adjust_auto_exposure_time` will start for the a previously determined exposure time and perform an adjustment. Something like, `determine` runs a loop and only returns when it gets a good value and `adjust` runs 1 iteration of the loop. 

In this method you would start by calling `determine`, before entering the loop. Once inside the loop, it would only call `adjust`. 

I have some comments about how you are determining this exposure time but I'll leave them for later, once we break this down.

Reply:
Done though maybe not exactly as you had in mind.